### PR TITLE
KEYCLOAK-16548 Collect more Keycloak metrics

### PR DESCRIPTION
**Pull request by slaskawi**
## JIRA ID

https://issues.redhat.com/browse/KEYCLOAK-16548

## Additional Information

This Pull Request creates additional Service for gathering Wildfly metrics. However, the most important change is to the Dashboard. Here's how it looks now [link](https://snapshot.raintank.io/dashboard/snapshot/fcIxz4Iu6qwuVhI3EU3JcgUtS2bWmvrU). Note, that some graphs didn't render correctly because RHSSO doesn't provide any way to enable metrics (we probably need to implement it - checking with @iankko).

image

## Checklist:
- [ ] Automated Tests
- [ ] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary
- [ ] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

### CODE REVIEW

**Comment by iankko** [comment]
@slaskawi 

> Note, that some graphs didn't render correctly because RHSSO doesn't provide any way to enable metrics (we probably need to implement it - checking with @iankko).

See the email reply (IMHO it's just question of choosing an option, which best suits your purpose):
> If there's no dedicated option - is there an option to break into the container in a nice way? :) I can think of some dirty tricks as replacing the bootstrap script etc. But that's not nice, and complicated from the upgrade perspective...
```
Looked, but there doesn't seem to be a dedicated option available so far, AFAICT. They could be enabled via three ways AFAICT.
Listing them in order of difficulty / time demands necessary to implement them:

1) Using CLI extensions - if it's enough just to switch the "statistics-enabled" property on some datasource from "false" to "true" and this is doable via JBoss CLI, then it should be feasible to achieve via extensions.cli (but didn't try):

https://github.com/jboss-container-images/redhat-sso-7-openshift-image/blob/sso74-dev/modules/sso/sso-extensions-cli/added/configure_sso_cli_extensions.sh#L7

2) Second way is to define custom definition of the datasource in question (but more complex IMHO than 1) ):

https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.3/html-single/getting_started_with_jboss_eap_for_openshift_online/index#building-custom-layers_default

(scroll down to "postgresql-datasource" definition)

3) If none of the above would work, we could add a dedicated option to have it enabled. Seems we are still using a dedicated
(overridden datasource code), derived from EAP:

https://github.com/jboss-container-images/redhat-sso-7-openshift-image/blob/sso74-dev/modules/sso/config/launch/setup/74/added/launch/datasource.sh#L8

Thus could add the new option / variable there. But this would take the longest time to achieve (implement + PR review
+ time till this gets into the next RH-SSO release).

Hope (at least some of the above) helps
Jan
```

**Comment by slaskawi** [comment]
Thanks for the explanation @iankko. We probably need to go with option number 3.

**Comment by pedroigor** [comment]
@slaskawi I remember that one issue we had was related to the `Number of entries in caches` not showing the information. I remember we discussing about issues in ISPN subsystem statistics/metrics ... Not sure.

/cc @sguilhen

**Comment by david-martin** [comment]
👍  @slaskawi This looks great.
1 question about the HTTP Errors panel.
How does it look when there are a lot of different routes and codes to sum by?
Will the route contain any uuids?

**Comment by Rajagopalan-Ranganathan** [comment]
@slaskawi : Can the panels be rearranged so that we don't have many empty spaces. Also, should try to rearrange them with the thought of which of those metrics/panels are of great importance to SRE. (Cosmetic change, but the goal is minimize scrolling and SRE people getting the important information at the top of their screen).

**Comment by Rajagopalan-Ranganathan** [comment]
@slaskawi a Follow up issue/task could be to create more specific dashboards that can be landed from a drill-down link in the main dashboard (this would keep the top level dashboard simple and SRE can then visit the specific dashboard for more details). Ex: Aggregated use of compute metrics shown in landing dashboard and drill-down link takes to specific pods/namespaces/ e.t.c

File: pkg/controller/keycloak/keycloak_reconciler_test.go [hunk 1]
```diff
@@ -54,11 +54,11 @@ func TestKeycloakReconciler_Test_Creating_All(t *testing.T) {
 	//    7) Postgresql Service
 	//    8) Keycloak Service
 	//    9) Keycloak Discovery Service
-	//    10) Keycloak Probe ConfigMap
-	//    11) Keycloak StatefulSets
-	//    12) Keycloak Route
+	//    10) Keycloak Monitoring Service
+	//    11) Keycloak Probe ConfigMap
+	//    12) Keycloak StatefulSets
 	//        Note, that's no MetricsRoute as it needs an established hostname from the root route
```
Comment (Rajagopalan-Ranganathan) [inline pkg/controller/keycloak/keycloak_reconciler_test.go:60]: Comment the 13th item here.

File: pkg/controller/keycloak/keycloak_reconciler_test.go [hunk 2]
```diff
@@ -54,11 +54,11 @@ func TestKeycloakReconciler_Test_Creating_All(t *testing.T) {
 	//    7) Postgresql Service
 	//    8) Keycloak Service
 	//    9) Keycloak Discovery Service
-	//    10) Keycloak Probe ConfigMap
-	//    11) Keycloak StatefulSets
-	//    12) Keycloak Route
+	//    10) Keycloak Monitoring Service
+	//    11) Keycloak Probe ConfigMap
+	//    12) Keycloak StatefulSets
```
Comment (Rajagopalan-Ranganathan) [inline pkg/controller/keycloak/keycloak_reconciler_test.go:59]: Can we add the comment for the newly added 13th (14th ) item here?

File: pkg/controller/keycloak/keycloak_reconciler_test.go [hunk 3]
```diff
@@ -54,11 +54,11 @@ func TestKeycloakReconciler_Test_Creating_All(t *testing.T) {
...
 	//        Note, that's no MetricsRoute as it needs an established hostname from the root route
-	assert.Equal(t, len(desiredState), 13)
+	assert.Equal(t, len(desiredState), 14)
```
Comment (Rajagopalan-Ranganathan) [inline pkg/controller/keycloak/keycloak_reconciler_test.go:61]: Would be nice to have the length of the dseried state defined as a constant. Easier to read, change and maintainable. Single line change in the future.

**Comment by slaskawi** [comment]
> Can the panels be rearranged so that we don't have many empty spaces. Also, should try to rearrange them with the thought of which of those metrics/panels are of great importance to SRE. (Cosmetic change, but the goal is minimize scrolling and SRE people getting the important information at the top of their screen).

@Rajagopalan-Ranganathan The panels are design to reside in one of the 3 sections: System health, Detailed Metrics and Experimental. The key design principles are exactly those as you mention - the System health panel enables the SRE Team to quickly tell if the service is healthy or not. The detailed metrics enables further diagnosis. The dashboard fits into a larger [Observability Proposal](https://github.com/keycloak/keycloak-community/pull/177).

So to sum it up - being an SRE Team you should probably never scroll the dashboard down. Being a developer, who investigates the root cause - you will probably need to go though most of the panels.

> a Follow up issue/task could be to create more specific dashboards that can be landed from a drill-down link in the main dashboard (this would keep the top level dashboard simple and SRE can then visit the specific dashboard for more details). Ex: Aggregated use of compute metrics shown in landing dashboard and drill-down link takes to specific pods/namespaces/ e.t.c

@Rajagopalan-Ranganathan So what exactly is missing? Perhaps this is something we could discuss first in the Observability proposal? https://github.com/keycloak/keycloak-community/pull/177

**Comment by Rajagopalan-Ranganathan** [comment]
> > Can the panels be rearranged so that we don't have many empty spaces. Also, should try to rearrange them with the thought of which of those metrics/panels are of great importance to SRE. (Cosmetic change, but the goal is minimize scrolling and SRE people getting the important information at the top of their screen).
> 
> @Rajagopalan-Ranganathan The panels are design to reside in one of the 3 sections: System health, Detailed Metrics and Experimental. The key design principles are exactly those as you mention - the System health panel enables the SRE Team to quickly tell if the service is healthy or not. The detailed metrics enables further diagnosis. The dashboard fits into a larger [Observability Proposal](https://github.com/keycloak/keycloak-community/pull/177).
> 
> So to sum it up - being an SRE Team you should probably never scroll the dashboard down. Being a developer, who investigates the root cause - you will probably need to go though most of the panels.
> 
> > a Follow up issue/task could be to create more specific dashboards that can be landed from a drill-down link in the main dashboard (this would keep the top level dashboard simple and SRE can then visit the specific dashboard for more details). Ex: Aggregated use of compute metrics shown in landing dashboard and drill-down link takes to specific pods/namespaces/ e.t.c
> 
> @Rajagopalan-Ranganathan So what exactly is missing? Perhaps this is something we could discuss first in the Observability proposal? [keycloak/keycloak-community#177](https://github.com/keycloak/keycloak-community/pull/177)
@slaskawi My thought was to may be move some of the panels ( like the ones the developer is interested and not the SRE) and have them as separate dashboards so that a deep link from the landing/main dashboard would take us to more detailed metrics dashboards. this would keep the main dashboard simple for the SRE.

**Comment by slaskawi** [comment]
> My thought was to may be move some of the panels ( like the ones the developer is interested and not the SRE) and have them as separate dashboards so that a deep link from the landing/main dashboard would take us to more detailed metrics dashboards. this would keep the main dashboard simple for the SRE.

@Rajagopalan-Ranganathan ooohhh no - this might not be a good idea! Please remember that one of the most powerful features is to limit the time range (or "zoom the dashboard in") and deeply investigate what happened in a specific period of time. Then you definitely need information about readiness probes, latency etc on a single dashboard and you definitely don't want to jump between the multiple ones.

Just to give you a specific example, if your memory gets full, you will first notice an increased overall latency in the System health panel (the top one). Then you can zoom the dashboard in and move down noticing higher CPU usage and higher GC rate. Those are intentionally next to each other. Finally, you move your eyes left to the memory usage and you have full information about what's going on.

Maybe we could tweak placement for some of those graphs but it requires some thought. In general the approach is: go top to bottom and left to right to gather details on what's going on.

File: pkg/controller/keycloak/keycloak_reconciler_test.go (hunk 2 above)
Comment (slaskawi) [reply pkg/controller/keycloak/keycloak_reconciler_test.go:59]: Done.

File: pkg/controller/keycloak/keycloak_reconciler_test.go (hunk 3 above)
Comment (slaskawi) [reply pkg/controller/keycloak/keycloak_reconciler_test.go:61]: This is one of the problems that we have with no good solution to be honest. 

On one hand, we decided to test the reconciler using unit tests. That decision turned out to work really great. On the other hand, we need additional "safety net" to verify if the reconciler creates all the objects we need in proper order. An alternative would be to use integration tests for, which is even worse.

So to get back to your suggestion - I don't think using a variable in this case makes sense. The variable would be used in only one place - to verify if the number of emitted objects is fine.

Comment (Rajagopalan-Ranganathan) [reply pkg/controller/keycloak/keycloak_reconciler_test.go:62]: @slaskawi After a quick look, a top of my mind suggestion would be, create a method to generate/give a `common.DesiredClusterState` that would be expected for the test case. And then compare the expected object (returned by the method) to the one returned by the reconciler. This should eliminate the need for index by index assertion of expected things. Would this be OK? Am i missing something?

Comment (slaskawi) [reply pkg/controller/keycloak/keycloak_reconciler_test.go:62]: That's a neat idea. Let me create a JIRA for this.

**Comment by Rajagopalan-Ranganathan** [comment]
> > My thought was to may be move some of the panels ( like the ones the developer is interested and not the SRE) and have them as separate dashboards so that a deep link from the landing/main dashboard would take us to more detailed metrics dashboards. this would keep the main dashboard simple for the SRE.
> 
> @Rajagopalan-Ranganathan ooohhh no - this might not be a good idea! Please remember that one of the most powerful features is to limit the time range (or "zoom the dashboard in") and deeply investigate what happened in a specific period of time. Then you definitely need information about readiness probes, latency etc on a single dashboard and you definitely don't want to jump between the multiple ones.
> 
> Just to give you a specific example, if your memory gets full, you will first notice an increased overall latency in the System health panel (the top one). Then you can zoom the dashboard in and move down noticing higher CPU usage and higher GC rate. Those are intentionally next to each other. Finally, you move your eyes left to the memory usage and you have full information about what's going on.
> 
> Maybe we could tweak placement for some of those graphs but it requires some thought. In general the approach is: go top to bottom and left to right to gather details on what's going on.

Thanks @slaskawi it makes sense now. We can rearrange the panels after receiving feedback from the AppSRE people as well. I am sure they have some thoughts on this.

File: pkg/controller/keycloak/keycloak_reconciler_test.go (hunk 3 above)
Comment (Rajagopalan-Ranganathan) [reply pkg/controller/keycloak/keycloak_reconciler_test.go:62]: @slaskawi thanks a lot!

File: pkg/common/auto_detect.go [hunk 4]
```diff
@@ -1,8 +1,6 @@
 package common
 
 import (
-	"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (vmuzikar) [inline pkg/common/auto_detect.go:4]: Just out of curiosity, why are we removing this package across the project?

File: pkg/model/keycloak_deployment.go [hunk 5]
```diff
@@ -152,6 +152,10 @@ func getKeycloakEnv(cr *v1alpha1.Keycloak, dbSecret *v1.Secret) []v1.EnvVar {
 			Name:  "PROXY_ADDRESS_FORWARDING",
 			Value: "true",
 		},
+		{
+			Name:  "KEYCLOAK_STATISTICS",
```
Comment (vmuzikar) [inline pkg/model/keycloak_deployment.go:156]: Is this variable documented somewhere? I suppose it's a var used by KC, not by the operator.

**Review by vmuzikar — APPROVED** [review]
As far as I can tell (this is a bit out of scope of my expertise :)), looks good to me but I didn't try it. Added two questions.

File: pkg/common/auto_detect.go (hunk 4 above)
Comment (slaskawi) [reply pkg/common/auto_detect.go:4]: I simplified the auto detecting mechanism. It turned out it contained a race condition (between the reconciler and the channel in the `auto_detect.go`). Once I did that, the schema started to be useless. 

I plan to simplify it a bit more in the future. I'd like to squash Auto Detect mechanism with the Background object we have. Currently, it's hard to wrap your head around it...

File: pkg/model/keycloak_deployment.go (hunk 5 above)
Comment (slaskawi) [reply pkg/model/keycloak_deployment.go:156]: Here's the manual for this: https://github.com/keycloak/keycloak-containers/tree/master/server#enable-some-metrics

In the Operator, we need the metrics for Grafana.

**Review by pedroigor — DISMISSED** [review]

**Review by pedroigor — APPROVED** [review]
